Reliability and Calibrated Prediction

NathanielF opened the pull request:
This will be an example notebook demonstrating the techniques of survival analysis used in reliability context with a focus on the predictive distribution and calibrated prediction intervals from a Bayesian and MLE perspective. Relates to issue: https://github.com/pymc-devs/pymc-examples/issues/474

Adding this draft version here, because i'm having a little trouble with the pre-commit checks

+ [X] Notebook follows style guide https://docs.pymc.io/en/latest/contributing/jupyter_style.html
+ [X] PR description contains a link to the relevant issue:
  * a tracker one for existing notebooks (tracker issues have the "tracker id" label)
  * or a proposal one for new notebooks
+ [x] Check the notebook is not excluded from any pre-commit check: https://github.com/pymc-devs/pymc-examples/blob/main/.pre-commit-config.yaml

NathanielF commented:
Hi @drbenvincent and @OriolAbril 

I think this is ready for review now. It's longer than i anticipated and i've cut out the idea to compare against conformal prediction as there is already quite a bit of detail in there.

The broader structure has three parts: 

<img width="700" alt="image" src="https://user-images.githubusercontent.com/10588167/211019384-22a2138d-7c4f-4a0d-bdc1-762fa86e84eb.png">

In the first part i'm just introducing time-to-failure data and the descriptive statistical analogues of survival and cdf functions. I follow a trajectory of first showing how we can analyse this kind of data with MLE or bootstrap style inference and then show on a trickier example with sparser data why the ability to add Bayesian priors to help calibrate the risk the profile is useful. Then i conclude with a "lets all be friends" kind of ending. 

I'm not sure who is best placed to review, but since i deal with censored data and I know @drbenvincent recently updated the censored data notebook, i thought you might find it interesting? Curious about what you think too @OriolAbril? Happy to take any feedback on the content or structure...

OriolAbril commented on examples/case_studies/reliability_and_calibrated_prediction.myst.md:
````diff
@@ -0,0 +1,1191 @@
+We'll use the posterior predictive distribution of the uniformative model. We show here how to derive the uncertainty in the estimates of the 95% prediction interval for the number of failures in a time interval.
+
+```{code-cell} ipython3
+def PI_failures(joint_draws, lp, up, n_at_risk):
````
I'll update the code to use xarray-einstats, the inputs are xarray dataset and scalars IIUC, so there should be no need for looping in order to use scipy distributions.

NathanielF replied:
On this one... i think i kind of disagree. While it's not the most efficient way to do this operation, i wanted the parallel with the bootstrapping looping procedures to be clear. We're instantiating in a loop a reasonable model fit to the data in each iteration of the loop, and seeing the variation induced in downstream statistics... i think pedagogically this is simpler than hiding the similarity of the procedure in more efficient code....

NathanielF replied:
Oops I think I misread this comment yesterday, did you want to try and streamline the loop? My only point was that I'd like to keep the analogy with the bootstrap procedure clear....

But I think I could just add more text explaining that too.

NathanielF replied:
Ok, i've added some more text to highlight the relationship between the bootstrap and posterior predictive distributions if you want to work some xarray magic, that'd be cool! Thanks!

OriolAbril replied:
Updated, and found a bug in xarray-einstats while doing it 😄 (I have fixed it already)

NathanielF replied:
Is it this one: 

<img width="763" alt="image" src="https://user-images.githubusercontent.com/10588167/212644548-dfb57b40-15d3-436e-b1b0-ae77cfb0c948.png">

I updated to my xarray einstats version, but still getting an error when i try to re-run your new function?

NathanielF replied:
Can recreate with just this call: 
<img width="690" alt="image" src="https://user-images.githubusercontent.com/10588167/212649410-32679f03-5fb6-4665-8c8c-95f8e788c5a6.png">

NathanielF replied:
Oh!!!! Sorry, i see the issue, the latest change is not on pip!!!! Sorry, slow this morning. 

<img width="362" alt="image" src="https://user-images.githubusercontent.com/10588167/212658370-fc1f356f-9b69-4aaa-a1b1-fbd0278a727c.png">

Ehm, that's cool. I'm happy with the changes and you can merge if you want. But for reproducibility should we be concerned that the user can't easily install the required packages?

NathanielF replied:
OK, cool! Let's do it. Thanks again for all your help on this one! I'm happy with it now if you are.

OriolAbril commented on examples/case_studies/reliability_and_calibrated_prediction.myst.md:
```diff
@@ -0,0 +1,1191 @@
+
+The choice of model in such cases is crucial. The decision about which failure profile is apt has to be informed by a subject matter expert because extrapolation from such sparse data is always risky. An understanding of the uncertainty is crucial if real costs attach to the failures and the subject matter expert is usually better placed to tell if you 2 or 7 failures can be plausibly expected wi ...
+
+# Conclusion
```
I think this should be level 2 header, otherwise the table of contents looks a bit weird: https://pymcio--491.org.readthedocs.build/projects/examples/en/491/case_studies/reliability_and_calibrated_prediction.html.

NathanielF replied:
Done, changed this.

OriolAbril commented on examples/case_studies/reliability_and_calibrated_prediction.myst.md:
```diff
@@ -0,0 +1,1191 @@
+
+## Authors
+
+Nathaniel Forde
```
This should follow the same format as other notebooks: https://docs.pymc.io/en/latest/contributing/jupyter_style.html#authorship-and-attribution

NathanielF replied:
Updated to better reflect the format.

OriolAbril reviewed (commented):
I have commented on the myst file, but the changes should be done in ipynb and then use pre-commit to update myst one.

General comment: I would use https://myst-nb.readthedocs.io/en/latest/render/hiding.html to hide the source of some of the code. i.e. the stringIO ones can be hidden, we would only need to split them so the `df_name` to show the rendered table is in its own cell right below (not hidden so readers can follow the notebook and get the variable name even without expanding the cell). Some of the plots I think could also have their source hidden.

NathanielF commented:
Thanks so much for the review. @OriolAbril. I've quibbled with one of the requests but i think i've addressed the rest. I've added hide-input tags to a bunch of the data loads and some of the more involved plotting functions.

OriolAbril commented on examples/case_studies/reliability_and_calibrated_prediction.myst.md:
```diff
@@ -358,9 +361,9 @@ def plot_cdfs(actuarial_table, dist_fits=True, ax=None, title="", xy=(3000, 0.5)
-        actuarial_table["t"],
-        actuarial_table["CI_95_lb"],
-        actuarial_table["CI_95_ub"],
+        actuarial_table["t"].values,
```
I had an error in matplotlib trying to run this, tried updating matplotlib and numpy and it didn't seem to fix the issue, so I added this values to continue ahead. I don't think it is a big issue though.

NathanielF replied:
Yeah, i think that's fine.

OriolAbril commented on examples/case_studies/reliability_and_calibrated_prediction.myst.md:
```diff
@@ -21,6 +21,9 @@ substitutions:
 :author: Nathaniel Forde
 :::
 
+:::{include} ../extra_installs.md
```
I noticed you had set the substitutions metadata but this include was missing. It should now add some auto instructions on installing lifelines.

NathanielF replied:
Oh, cool. Yeah, i forgot this,

OriolAbril commented on examples/case_studies/reliability_and_calibrated_prediction.myst.md:
````diff
@@ -444,23 +447,6 @@ def bayes_boot(df, lb, ub, seed=100):
 ```
 
 ```{code-cell} ipython3
-def bootstrap(df, lb, ub, seed=100):
````
This is the same function as above, and this one wasn't being used, so I deleted it.

NathanielF replied:
It was a slightly different function. One implemented the naive bootstrap and the other implemented the bayesian bootstrap with a flat dirichelet prior. But yeah, basically the same, so i'm fine getting rid of it.

OriolAbril commented on examples/case_studies/reliability_and_calibrated_prediction.myst.md:
```diff
@@ -953,15 +939,11 @@ az.summary(idata_informative)
-joint_draws = az.extract_dataset(idata, group="posterior", num_samples=1000)[
-    ["alpha", "beta"]
-].to_dataframe()
+joint_draws = az.extract(idata, num_samples=1000)[["alpha", "beta"]]
```
The fact that was a dataframe wasn't being used anywhere, so I changed that because now below we take advantage of the fact `joint_draws` is an xarray dataset.

NathanielF replied:
That's fair.

OriolAbril reviewed (commented):
I reran the notebook locally to update that sampling snippet. And did a couple extra changes that I have commented here in the review. Let me know what you think, they are all quite minor except for the sampling one.

NathanielF commented:
Thanks @OriolAbril I'm fine with all the changes you made, i understand the logic of the xarray-einstats version too, but i just couldn't get it to work on my machine and i'm unsure why? I've attached a screenshot above.

NathanielF commented:
Just fixing a typo at the beginning. Still have same xarray issue.

NathanielF commented:
Woop, woop! Thanks 😊!
